File: historian_v3/pkg/historian/retention.go
package historian

import (
	"os"
	"path/filepath"
	"strings"
	"time"
)

func StartRetentionLoop(cfg Config) {
	if !cfg.Retention.Enabled {
		return
	}
	ticker := time.NewTicker(time.Duration(cfg.Retention.CheckIntervalMs) * time.Millisecond)
	go func() {
		for range ticker.C {
			_ = runRetention(cfg)
		}
	}()
}

func runRetention(cfg Config) error {
	cutoff := time.Now().UTC().Add(-time.Duration(cfg.Retention.MaxAgeHours) * time.Hour)
	for _, root := range []string{"raw", "index"} {
		base := filepath.Join(cfg.Storage.DataDir, root)
		_ = filepath.Walk(base, func(path string, info os.FileInfo, err error) error {
			if err != nil || info == nil || !info.IsDir() {
				return nil
			}
			rel, err := filepath.Rel(base, path)
			if err != nil || rel == "." {
				return nil
			}
			parts := strings.Split(filepath.ToSlash(rel), "/")
			if len(parts) != 2 {
				return nil
			}
			t, err := time.Parse(time.RFC3339, parts[0]+"T"+parts[1]+":00:00Z")
			if err != nil {
				return nil
			}
			if t.Before(cutoff) {
				_ = os.RemoveAll(path)
			}
			return nil
		})
	}
	return nil
}

File: historian_v2/src/storage/retention.rs
use anyhow::Result;
use chrono::{DateTime, Utc};
use std::path::Path;
use tokio::time::{sleep, Duration};

use crate::config::Config;

async fn cleanup_once(cfg: &Config) -> Result<()> {
    if !cfg.retention.enabled {
        return Ok(());
    }
    let cutoff = Utc::now().timestamp_millis() - cfg.retention.max_age_hours * 3_600_000;
    for root in ["raw", "index"] {
        let top = Path::new(&cfg.storage.data_dir).join(root);
        if !top.exists() {
            continue;
        }
        let mut day_dir = tokio::fs::read_dir(&top).await?;
        while let Some(day) = day_dir.next_entry().await? {
            let day_name = day.file_name().to_string_lossy().to_string();
            let day_path = day.path();
            if !day_path.is_dir() {
                continue;
            }
            let mut hour_dir = tokio::fs::read_dir(&day_path).await?;
            while let Some(hour) = hour_dir.next_entry().await? {
                let hour_name = hour.file_name().to_string_lossy().to_string();
                let t = format!("{day_name}T{hour_name}:00:00Z");
                if let Ok(parsed) = DateTime::parse_from_rfc3339(&t.replace(' ', "T")) {
                    if parsed.with_timezone(&Utc).timestamp_millis() < cutoff {
                        let _ = tokio::fs::remove_dir_all(hour.path()).await;
                    }
                }
            }
        }
    }
    Ok(())
}

pub fn spawn_retention_task(cfg: Config) {
    tokio::spawn(async move {
        loop {
            let _ = cleanup_once(&cfg).await;
            sleep(Duration::from_millis(cfg.retention.check_interval_ms)).await;
        }
    });
}
